source/data-storage.lisp

```
@@ -0,0 +1,190 @@
+(defclass download-data-path (data-path) ())
+
+(defvar +data-root+ "next")
+
+(declaim (ftype (function (data-path &key (:root trivial-types:pathname-designator))
+                          (or string null))
+                expand-default-path))
+(defun expand-default-path (path &key (root (uiop:xdg-data-home +data-root+)))
+  "Derive file from command line option or NAME.
+When PATH has no basename, return ROOT.  This is useful to refer to directories.
+If NAME has a slash, return NAME.
+Without slash, NAME is expanded to 'ROOT/NAME.lisp' or 'ROOT/NAME' is NAME
+already contains a period."
+  ;; TODO: Get command line option.
+  ;; (setf name (or (getf *options* :with-path ...) name))
+  (with-slots (basename) path
+    (cond
+      ((uiop:emptyp basename)
+       root)
+      ((search "/" basename)
+       basename)
+      (t
+       (let ((name (str:concat (namestring root) "/" basename)))
+         (unless (search "." basename)
+           (setf name (str:concat name ".lisp")))
+         name)))))
+
+(defmethod expand-data-path ((path data-path) (profile data-profile))
+  "Return finalized path.
+Return NIL when path must not be used.  This makes it possible to use the
+function result as a boolean in conditions."
+  (expand-default-path path))
+
+(defmethod expand-data-path ((path init-file-data-path) (profile data-profile))
+  "Return finalized path.
+Return NIL when path must not be used.  This makes it possible to use the
+function result as a boolean in conditions."
+  (expand-default-path path :root (uiop:xdg-config-home +data-root+)))
+
+(defmethod expand-data-path ((path session-data-path) (profile data-profile))
+  (expand-default-path path :root (uiop:xdg-data-home +data-root+ "sessions")))
+
+(defun xdg-download-dir ()
+  (let ((dir (ignore-errors (uiop:run-program '("xdg-user-dir" "DOWNLOAD")
+                                              :output '(:string :stripped t)))))
+    (when (or (null dir) (string= dir (uiop:getenv "HOME")))
```

could just check return code of uiop:run-program, I believe one of the values is the return code

---

Yes but why?  We don't need the return code.
I'm ignoring errors mostly in case the `xdg-user-dir` executable is not on the system.

---

return code will tell you whether the dir exists without having to check `(null dir)`

---

It does not cover the case when `dir` is nil because `xdg-user-dir` is missing.